x/accounts/vesting/periodic_vesting_account.go

```diff
@@ -0,0 +1,257 @@
+package vesting
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"cosmossdk.io/collections"
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	"cosmossdk.io/x/accounts/accountstd"
+	account_abstractionv1 "cosmossdk.io/x/accounts/interfaces/account_abstraction/v1"
+	vestingtypes "cosmossdk.io/x/accounts/vesting/types"
+	banktypes "cosmossdk.io/x/bank/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// Compile-time type assertions
+var (
+	_ accountstd.Interface = (*PeriodicVestingAccount)(nil)
+)
+
+// NewPeriodicVestingAccount creates a new PeriodicVestingAccount object.
+func NewPeriodicVestingAccount(d accountstd.Dependencies) (*PeriodicVestingAccount, error) {
+	baseVesting := NewBaseVesting(d)
+
+	periodicsVestingAccount := PeriodicVestingAccount{
+		BaseVesting:    baseVesting,
+		StartTime:      collections.NewItem(d.SchemaBuilder, StartTimePrefix, "start_time", sdk.IntValue),
+		VestingPeriods: collections.NewMap(d.SchemaBuilder, VestingPeriodsPrefix, "vesting_periods", collections.StringKey, codec.CollValue[vestingtypes.Period](d.LegacyStateCodec)),
+	}
+
+	return &periodicsVestingAccount, nil
+}
+
+type PeriodicVestingAccount struct {
+	*BaseVesting
+	StartTime      collections.Item[math.Int]
+	VestingPeriods collections.Map[string, vestingtypes.Period]
+}
+
+func (pva PeriodicVestingAccount) Init(ctx context.Context, msg *vestingtypes.MsgInitPeriodicVestingAccount) (*vestingtypes.MsgInitPeriodicVestingAccountResponse, error) {
+	sender := accountstd.Sender(ctx)
+	if sender == nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("Cannot find sender address from context")
+	}
+	to := accountstd.Whoami(ctx)
+	if to == nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("Cannot find account address from context")
+	}
+
+	toAddress, err := pva.addressCodec.BytesToString(to)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid 'to' address: %s", err)
+	}
+	fromAddress, err := pva.addressCodec.BytesToString(sender)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid 'to' address: %s", err)
+	}
+
+	if msg.StartTime < 1 {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "invalid start time of %d, length must be greater than 0", msg.StartTime)
+	}
+
+	totalCoins := sdk.Coins{}
+	endTime := msg.StartTime
+	for i, period := range msg.VestingPeriods {
+		if period.Length < 1 {
+			return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "invalid period length of %d in period %d, length must be greater than 0", period.Length, i)
+		}
+
+		if err := validateAmount(period.Amount); err != nil {
+			return nil, err
+		}
+
+		totalCoins = totalCoins.Add(period.Amount...)
+		// Calculate end time
+		endTime += period.Length
+		err = pva.VestingPeriods.Set(ctx, fmt.Sprint(i), period)
```

Comment:
why are we fmt.Sprint i into string? 

in the collection we could map it, or if u want to treat is as a slice u can use collections.Vec[Period]; although this feels like should be a period end time, which should be computed when the vesting account is created

Reply:
ah thanks, Vec fit this perfectly